### PR TITLE
Always generate checksums as last part of publish job

### PR DESCRIPTION
**Pull request by wtgodbe**
My first iteration of publishing checksums for this repo didn't work, since it generated checksums for some assets, _before_ those assets were signed (specifically the installer zips). Instead, I'm here trying to ensure that checksums are only generated once per build, after all build and signing jobs are done. I'm using the assumption that the step that passes `_PublishArgs` is always the final build step, and therefore we can safely hook the checksum target into `AfterSigning.proj` in that leg (note that only the winx64/x86 job has multiple build steps - all the other jobs just run `build` once & pass `_PublishArgs` to that).

This should generate checksums for the same set of assets as my previous solution.

Internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=581767

Resolves part of https://github.com/dotnet/aspnetcore/issues/18792 (still need to port this to master, and find another solution for 2.1)

### CODE REVIEW

File: eng/AfterSigning.targets [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<Project>
```
Comment (wtgodbe) [inline eng/AfterSigning.targets]: Project gets imported here: https://github.com/dotnet/arcade/blob/cc4af3898310cf2826d78de453dff4a25905b888/src/Microsoft.DotNet.Arcade.Sdk/tools/AfterSigning.proj#L29

**Comment by wtgodbe** [comment]
Initial build didn't work since `AfterSigning.targets` doesn't import root `d.b.props`, so it didn't have some properties I was expecting it to. New build: https://dev.azure.com/dnceng/internal/_build/results?buildId=581882

**Comment by wtgodbe** [comment]
Last build didn't work either, because `AfterSigning.proj` in Arcade doesn't import the Arcade SDK props/targets files. New build: https://dev.azure.com/dnceng/internal/_build/results?buildId=582051

**Comment by wtgodbe** [comment]
New build with a tailored list of extensions to generate checksums for: https://dev.azure.com/dnceng/internal/_build/results?buildId=582357

**Review by mthalman — APPROVED** [review]
I'm no expert on the product build side of things but things seem fine to me.

**Comment by wtgodbe** [comment]
snipster

Here's the set of checksums that got published - I'm working on verifying that they're valid. @dougbu @JunTaoLuo @mthalman any last concerns with the content of the PR, or the content of what it generated?

File: src/Installers/Windows/Wix.targets [hunk 2]
```diff
@@ -64,18 +64,12 @@
 
   <Target Name="CopyToArtifactsDirectory"
           Condition=" '$(IsShipping)' == 'true' AND '$(SkipCopyToArtifactsDirectory)' != 'true' "
-          AfterTargets="Build" BeforeTargets="GenerateChecksums">
+          AfterTargets="Build">
```
Comment (dougbu) [inline src/Installers/Windows/Wix.targets:67]: Use `BeforeTargets="Build"` because `Build` is a no-op target that comes at the _end_ of the build process. As-is, there's no guarantee when this target will run. (Actually, it'll likely run when the calling project is about to exit and that's long before AfterSigning.targets would execute. Still good practice.)

**Review by dougbu — APPROVED** [review]
A few nits and questions but only one "should fix". I don't need to review again though I am interested in the questions.

**Comment by wtgodbe** [comment]
I validated that the checksums are correct this time (w/ `$(Get-FileHash <filepath> -Algorithm SHA512).Hash`, but noticed that I missed the .rpm and .wixlib files. I'll add those.

File: eng/AfterSigning.targets [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  <Import Project="Sdk.props" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(GenerateChecksums)' == 'true'" />
+  <!-- $(InstallersOutputPath) is not defined. Root Directory.Build.props is not imported. -->
+
+  <PropertyGroup>
+    <!-- The one use of ArtifactsDir in Publish.proj adds an additional slash, confusing itself. -->
+    <ArtifactsDir Condition=" HasTrailingSlash('$(ArtifactsDir)') ">$(ArtifactsDir.Substring(0, $([MSBuild]::Subtract($(ArtifactsDir.Length), 1))))</ArtifactsDir>
```
[the remark this answers is not in the document]
Comment (wtgodbe) [reply eng/AfterSigning.targets:8]: I took this from publishing.props in case it was also needed here - I'm not actually sure if it is. They aren't loaded in the same context.

File: eng/Publishing.props [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <Project>
-  <PropertyGroup Condition=" HasTrailingSlash('$(ArtifactsDir)') ">
+  <PropertyGroup>
     <!-- The one use of ArtifactsDir in Publish.proj adds an additional slash, confusing itself. -->
-    <ArtifactsDir>$(ArtifactsDir.Substring(0, $([MSBuild]::Subtract($(ArtifactsDir.Length), 1))))</ArtifactsDir>
+    <ArtifactsDir Condition=" HasTrailingSlash('$(ArtifactsDir)') ">$(ArtifactsDir.Substring(0, $([MSBuild]::Subtract($(ArtifactsDir.Length), 1))))</ArtifactsDir>
```
[the remark this answers is not in the document]
Comment (wtgodbe) [reply eng/Publishing.props:4]: I moved this down here so it wouldn't affect the other properties in the propertyGroup (e.g. `PublishDependsOnTargets` wouldn't get set if `ArtifactsDir` had a trailing slash).

File: src/Installers/Windows/Wix.targets (hunk 2 above)
Comment (wtgodbe) [reply src/Installers/Windows/Wix.targets:67]: This is an undo of a change I made in my first PR to generate checksums - it's intended to no-op to where we were before I added checksum stuff to the repo. This project should now do nothing checksum-related. See:

https://github.com/dotnet/aspnetcore/blob/03b4bcfee8a2e8de292e3bbfceaca9d58b60e92e/src/Installers/Windows/Wix.targets#L65-L75

**Comment by wtgodbe** [comment]
New build which should include wixlib/rpm checksums: https://dev.azure.com/dnceng/internal/_build/results?buildId=583841

File: src/Installers/Windows/Wix.targets (hunk 2 above)
Comment (dougbu) [reply src/Installers/Windows/Wix.targets:67]: That's fine but `BeforeTargets` pointing to a no-op target (like `Build`) is _always_ more deterministic and reliable than any `AfterTargets`

**Comment by wtgodbe** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=584014

File: eng/Publishing.props (hunk 4 above)
[the remark this answers is not in the document]
Comment (dougbu) [reply eng/Publishing.props:4]: 👍

**Review by dougbu — APPROVED** [review]

**Comment by wtgodbe** [comment]
Yarn endpoint exploded. I'll hit retry when I can

File: src/Installers/Windows/Wix.targets [hunk 5]
```diff
@@ -64,18 +64,12 @@
 
   <Target Name="CopyToArtifactsDirectory"
           Condition=" '$(IsShipping)' == 'true' AND '$(SkipCopyToArtifactsDirectory)' != 'true' "
-          AfterTargets="Build" BeforeTargets="GenerateChecksums">
```
Comment (JunTaoLuo) [inline src/Installers/Windows/Wix.targets:67]: Why is this changed?

**Review by JunTaoLuo — APPROVED** [review]
One question but otherwise looks good.

File: src/Installers/Windows/Wix.targets (hunk 5 above)
Comment (wtgodbe) [reply src/Installers/Windows/Wix.targets:67]: This is back to the way the file was before my first checksum change, except with 'AfterTargets=Build' changed to 'BeforeTargets=Build'. Doug suggested the change since `Build` is a no-op target that runs at the end of the project build, and it's more deterministic to hook into it w/ a `Before` than an `After`

Comment (JunTaoLuo) [reply src/Installers/Windows/Wix.targets:67]: Ah I missed that comment.
